### PR TITLE
Add `GetProcessedMetrics` method to Prometheus helper

### PR DESCRIPTION
**Pull request by exekias**
This PR adds a new `GetProcessedMetrics` method to ease the creation of metricsets based on Prometheus. It moves from code (current approach) to a mapping template for the conversion of Prometheus metrics to our format, better suited for Elasticsearch mappings.

With this new method you only need to fill the `MetricsMapping` structure to map all metrics to events. There are a couple of different use cases to support different patterns in Prometheus metrics, like a way to extract labels into `keyworud` values, or doing boolean conversion.

This PR is divided into 2 commits: 
1. Introducing the method
2. Adapt existing `state_*` metricsets to make use of it

I plan to use this for incoming modules/metricsets.

:warning: There is a breaking change affecting the `state_container` metricset (currently beta):

- Fixed a typo in `status.phase` value: `terminate` -> `terminated`

Also this deprecates calculated metric `cpu.*.nanocores` in favor of `cpu.*.cores`.

### CODE REVIEW

File: metricbeat/helper/prometheus/label.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+// Label maps a Prometheus label to a Metricbeat field
+func Label(field string) LabelMap {
+	return &commonLabel{
+		field: field,
+		key:   false,
+	}
+}
+
+// KeyLabel maps a Prometheus label to a Metricbeat field. The lable is flagged as key.
```
Comment (jsoriano) [inline metricbeat/helper/prometheus/label.go]: Typo `lable`

File: metricbeat/module/kubernetes/state_container/_meta/fields.yml [hunk 2]
```diff
@@ -26,14 +26,14 @@
     - name: cpu
       type: group
       fields:
-        - name: limit.nanocores
+        - name: limit.cores
```
Comment (ruflin) [inline metricbeat/module/kubernetes/state_container/_meta/fields.yml]: As `state_container` is already GA we should not do breaking changes. Could we perhaps keep both fields but mark one as deprecated? Same below.

File: metricbeat/helper/prometheus/prometheus.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	Metrics map[string]MetricMap
+
+	// Labels translate from prometheus label names to Metricbeat fields
+	Labels map[string]LabelMap
+
+	// ExtraFields adds the given fields to all events coming from `GetProcessedMetrics`
+	ExtraFields map[string]string
+}
+
+func (p *prometheus) GetProcessedMetrics(mapping *MetricsMapping) ([]common.MapStr, error) {
```
Comment (ruflin) [inline metricbeat/helper/prometheus/prometheus.go]: It reminds me a lot of our schema package.

File: metricbeat/helper/prometheus/prometheus.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	eventsMap := map[string]common.MapStr{}
+	for _, family := range families {
+		for _, metric := range family.GetMetric() {
+			m, ok := mapping.Metrics[family.GetName()]
+
+			// Ignore unknown metrics
+			if !ok {
+				continue
```
Comment (ruflin) [inline metricbeat/helper/prometheus/prometheus.go]: We should sync up the behaviour with the schema package on what we do on error / missing fields. I'm not happy with the current error handling in schema. Let's have a follow up chat on how we should push that forward.

File: CHANGELOG.asciidoc [hunk 5]
```diff
@@ -39,6 +39,8 @@ https://github.com/elastic/beats/compare/v6.0.0-beta2...master[Check the HEAD di
 - Add config option for windows/perfmon metricset to ignore non existent counters. {pull}6432[6432]
 - Refactor docker CPU calculations to be more consistent with `docker stats`. {pull}6608[6608]
 - Update logstash.node_stats metricset to write data under `logstash.node.stats.*`. {pull}6714[6714]
+- Changed `state_container` metric from `cpu.*.nanocores` to `cpu.*.cores`.
+- Fixed typo in values for `state_container` `status.phase`, from `terminate` to `terminated`.
```
Comment (ruflin) [inline CHANGELOG.asciidoc]: Here we could perhaps use `copy_to` to copy the field for now?

File: metricbeat/helper/prometheus/prometheus.go (hunk 3 above)
Comment (exekias) [reply metricbeat/helper/prometheus/prometheus.go]: Definitely inspired on it :)

File: metricbeat/helper/prometheus/metric.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+type MetricMap interface {
+	// GetField returns the resulting field name
+	GetField() string
+
+	// GetValue returns the resulting value
+	GetValue(m *dto.Metric) interface{}
+}
+
+// Metric directly maps a Prometheus metric to a Metricbeat field
+func Metric(field string) MetricMap {
```
[the remark this answers is not in the document]
Comment (exekias) [reply metricbeat/helper/prometheus/metric.go]: Uhm that's a good question, so far I think that mapping should take care of that?

File: metricbeat/module/kubernetes/state_container/_meta/fields.yml (hunk 2 above)
Comment (exekias) [reply metricbeat/module/kubernetes/state_container/_meta/fields.yml]: I've pushed a change keeping them (deprecated)

File: CHANGELOG.asciidoc (hunk 5 above)
Comment (exekias) [reply CHANGELOG.asciidoc]: I used code as there is a change on the scale, so needed to perform the math

File: metricbeat/helper/prometheus/metric.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ruflin) [reply metricbeat/helper/prometheus/metric.go]: It indeed does for the define fields. Coming from the `schema` package I kind of expected to have both.

**Pull request merged**